### PR TITLE
Add support for aws_cloudwatch_metric_stream

### LINKED ISSUE #18525: Amazon CloudWatch Metric Streams
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> Announcing Amazon CloudWatch Metric Streams - a new feature that enables customers to create a continuous, near real-time stream of metrics to a destination of their choice.

### New or Affected Resource(s)

* `aws_cloudwatch_metric_stream`

### Potential Terraform Configuration

```hcl
resource "aws_cloudwatch_metric_stream" "example" {
  name          = "example"
  role_arn      = aws_iam_role.example.arn
  firehose_arn  = aws_kinesis_firehose_delivery_stream.example.arn
  output_format = "json"
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2021/03/announcing-amazon-cloudwatch-metric-streams/).
[Blog post](https://aws.amazon.com/blogs/aws/cloudwatch-metric-streams-send-aws-metrics-to-partners-and-to-your-apps-in-real-time/).
[User Guide](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Metric-Streams.html).

### PR DESCRIPTION
**Pull request by gcacace**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #18525

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchMetricStream'

TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchMetricStream -timeout 180m
=== RUN   TestAccAWSCloudWatchMetricStream_basic
=== PAUSE TestAccAWSCloudWatchMetricStream_basic
=== RUN   TestAccAWSCloudWatchMetricStream_noName
=== PAUSE TestAccAWSCloudWatchMetricStream_noName
=== RUN   TestAccAWSCloudWatchMetricStream_namePrefix
=== PAUSE TestAccAWSCloudWatchMetricStream_namePrefix
=== RUN   TestAccAWSCloudWatchMetricStream_includeFilters
=== PAUSE TestAccAWSCloudWatchMetricStream_includeFilters
=== RUN   TestAccAWSCloudWatchMetricStream_excludeFilters
=== PAUSE TestAccAWSCloudWatchMetricStream_excludeFilters
=== RUN   TestAccAWSCloudWatchMetricStream_update
=== PAUSE TestAccAWSCloudWatchMetricStream_update
=== RUN   TestAccAWSCloudWatchMetricStream_updateName
=== PAUSE TestAccAWSCloudWatchMetricStream_updateName
=== CONT  TestAccAWSCloudWatchMetricStream_basic
=== CONT  TestAccAWSCloudWatchMetricStream_update
=== CONT  TestAccAWSCloudWatchMetricStream_excludeFilters
=== CONT  TestAccAWSCloudWatchMetricStream_namePrefix
=== CONT  TestAccAWSCloudWatchMetricStream_includeFilters
=== CONT  TestAccAWSCloudWatchMetricStream_noName
=== CONT  TestAccAWSCloudWatchMetricStream_updateName
--- PASS: TestAccAWSCloudWatchMetricStream_excludeFilters (35.26s)
--- PASS: TestAccAWSCloudWatchMetricStream_includeFilters (35.35s)
--- PASS: TestAccAWSCloudWatchMetricStream_noName (35.40s)
--- PASS: TestAccAWSCloudWatchMetricStream_namePrefix (37.61s)
--- PASS: TestAccAWSCloudWatchMetricStream_basic (111.76s)
--- PASS: TestAccAWSCloudWatchMetricStream_update (125.37s)
--- PASS: TestAccAWSCloudWatchMetricStream_updateName (191.73s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	193.508s
```

### CODE REVIEW

**Comment by YakDriver** [comment]
@gcacace After a quick scan, this PR is looking good. Overall, the neatness and completeness, with implementation, docs, tests, is great! Thank you for submitting it! I may point out a few minor items. Also, I encourage you to check the "allow edits by maintainers" box to facilitate getting this merged more quickly.

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  s3_configuration {
+    role_arn   = aws_iam_role.firehose_to_s3.arn
+    bucket_arn = aws_s3_bucket.bucket.arn
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:138]: Split args into two lists:

The following arguments are required:

* `firehose_arn` 
* ...

The following arguments are optional:

* `exclude_filter`...
* ...

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+* `firehose_arn` - (Required) The Amazon Resource Name (ARN) of the Amazon Kinesis Firehose delivery stream to use for this metric stream.
+* `include_filter` - (Optional) List of inclusive metric filters. If you specify this parameter, the stream sends only the metrics from the metric namespaces that you specify here. Conflicts with `exclude_filter`.
+* `name` - (Optional, Forces new resource) Friendly name of the metric stream. If omitted, Terraform will assign a random, unique name. Conflicts with `name_prefix`.
+* `name_prefix` - (Optional, Forces new resource) Creates a unique friendly name beginning with the specified prefix. Conflicts with `name`.
+* `role_arn` - (Required) The Amazon Resource Name (ARN) of the IAM role that this metric stream will use to access Amazon Kinesis Firehose resources.
+* `output_format` - (Required) The output format for the stream. For more information about metric stream output formats, see [Metric streams output formats](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats.html).
+* `tags` - (Optional) A map of tags to assign to the resource.
+
+### Nested fields
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:149]:
```suggestion
```

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+#### `exclude_filter`
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:151]:
```suggestion
### `exclude_filter`
```

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+* `namespace` - (Required) The name of the metric namespace in the filter.
+
+#### `include_filter`
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:155]:
```suggestion
### `include_filter`
```

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+* `namespace` - (Required) The name of the metric namespace in the filter.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The Amazon Resource Name (ARN) of the metric stream.
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:163]:
```suggestion
* `arn` - ARN of the metric stream.
```
Here and everywhere, fully spelled out is not necessary.

File: website/docs/r/cloudwatch_metric_stream.html.markdown [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+* `firehose_arn` - (Required) The Amazon Resource Name (ARN) of the Amazon Kinesis Firehose delivery stream to use for this metric stream.
+* `include_filter` - (Optional) List of inclusive metric filters. If you specify this parameter, the stream sends only the metrics from the metric namespaces that you specify here. Conflicts with `exclude_filter`.
+* `name` - (Optional, Forces new resource) Friendly name of the metric stream. If omitted, Terraform will assign a random, unique name. Conflicts with `name_prefix`.
+* `name_prefix` - (Optional, Forces new resource) Creates a unique friendly name beginning with the specified prefix. Conflicts with `name`.
+* `role_arn` - (Required) The Amazon Resource Name (ARN) of the IAM role that this metric stream will use to access Amazon Kinesis Firehose resources.
+* `output_format` - (Required) The output format for the stream. For more information about metric stream output formats, see [Metric streams output formats](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats.html).
```
Comment (YakDriver) [inline website/docs/r/cloudwatch_metric_stream.html.markdown:146]:
```suggestion
* `output_format` - (Required) Output format for the stream. For more information about metric stream output formats, see [Metric streams output formats](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats.html).
```
Here and everywhere. Best practice is to drop the first article ("The", "A", "An) after either "(Optional)" or "(Required)".

**Review by YakDriver — CHANGES_REQUESTED** [review]
Thanks for you time on this!

File: aws/resource_aws_cloudwatch_metric_stream.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		if err := d.Set("exclude_filter", excludeFilters); err != nil {
+			return fmt.Errorf("error setting exclude_filter: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceAwsCloudWatchMetricStreamPut(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (gcacace) [reply aws/resource_aws_cloudwatch_metric_stream.go]: I can change the method to be `Create`. Would it then be ok to use that `Create` method for both `Create` and `Update` terraform actions? (Metric Streams don't have an Update API method, they use `PutMetricStream` to both create/update.

File: aws/resource_aws_cloudwatch_metric_stream.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		Read:   resourceAwsCloudWatchMetricStreamRead,
+		Update: resourceAwsCloudWatchMetricStreamPut,
+		Delete: resourceAwsCloudWatchMetricStreamDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
```
[the remark this answers is not in the document]
Comment (gcacace) [reply aws/resource_aws_cloudwatch_metric_stream.go:36]: You're welcome!

File: aws/resource_aws_cloudwatch_metric_stream.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+func resourceAwsCloudWatchMetricStreamRead(d *schema.ResourceData, meta interface{}) error {
+	name := d.Get("name").(string)
+	log.Printf("[DEBUG] Reading CloudWatch MetricStream: %s", name)
+	conn := meta.(*AWSClient).cloudwatchconn
+
+	params := cloudwatch.GetMetricStreamInput{
+		Name: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetMetricStream(&params)
```
[the remark this answers is not in the document]
Comment (gcacace) [reply aws/resource_aws_cloudwatch_metric_stream.go]: Done!

File: aws/resource_aws_cloudwatch_metric_stream.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+	d.Set("arn", resp.Arn)
+	d.Set("creation_date", resp.CreationDate.Format(time.RFC3339))
+	d.Set("firehose_arn", resp.FirehoseArn)
+	d.Set("last_update_date", resp.CreationDate.Format(time.RFC3339))
+	d.Set("name", resp.Name)
+	d.Set("output_format", resp.OutputFormat)
+	d.Set("role_arn", resp.RoleArn)
+	d.Set("state", resp.State)
+
+	if resp.IncludeFilters != nil && len(resp.IncludeFilters) > 0 {
```
[the remark this answers is not in the document]
Comment (gcacace) [reply aws/resource_aws_cloudwatch_metric_stream.go]: Done!

**Comment by gcacace** [comment]
Hi @YakDriver, wondering if I've missed anything here apart from your comments that I've addressed in the latest revision.

**Comment by frankreno** [comment]
Hi Folks, is there any eta on when this will be merged, released, and supported by TF? This is a great add we are ourselves waiting on.

**Comment by gcacace** [comment]
@YakDriver it has been almost 20 days since I've replied and fix all the review comments. Is there any chance we can get this merged?

**Comment by YakDriver** [comment]
@gcacace I was unable to push to your branch to fix remaining issues. I'm not sure if you "allow edits by maintainers" or if it fails because your branch name is "main." It is best to name branches following the convention of `f-` prefix for features, like a new resource (`d-` for docs, `td-` for technical debt, `b-` for bugs). One thing I notice that you'll need is to add, since the resource has tags, is the default tags capability. All resources should now have them. You can find examples in any resource with tags. Fixes needed for default tags:

1. `tags_all` addition to schema
2. Create: `defaultTagsConfig` addition, calculate `tags` var with `MergeTags()`, `tags.IgnoreAWS()...`
3. Read: `defaultTagsConfig`, `ignoreTagsConfig` additions, `if err := d.Set("tags"...`, `if err := d.Set("tags_all"...`
4. Update: `if d.HasChang("tags_all")...`

**Comment by YakDriver** [comment]
This will be superceded by #19429

**Pull request closed without merging**